### ISSUE TITLE
API: Create mechanism to allow adding new cards to existing note types

### ISSUE THREAD

**Issue by hanpingchinese** [ankidroid/Anki-Android#4889]
I've recently added (to my app) the ability to add a new card type via the AnkiDroid API. This works fine _unless_ I'm logged into my Anki account in AnkiDroid (2.8.3) settings, in which case I get:

```
E DatabaseUtils: Writing exception to parcel
E DatabaseUtils: java.lang.IllegalArgumentException: Unable to add template
E DatabaseUtils: 	at com.ichi2.anki.provider.CardContentProvider.insert(CardContentProvider.java:963)
E DatabaseUtils: 	at android.content.ContentProvider$Transport.insert(ContentProvider.java:267)
E DatabaseUtils: 	at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:152)
E DatabaseUtils: 	at android.os.Binder.execTransact(Binder.java:675)
E DatabaseUtils: Caused by: com.ichi2.anki.exception.ConfirmModSchemaException
E DatabaseUtils: 	at com.ichi2.libanki.Collection.modSchema(Collection.java:371)
E DatabaseUtils: 	at com.ichi2.libanki.Models.addTemplate(Models.java:762)
E DatabaseUtils: 	at com.ichi2.anki.provider.CardContentProvider.insert(CardContentProvider.java:958)
E DatabaseUtils: 	... 3 more
```

What can be causing this ConfirmModSchemaException?

**Comment by hanpingchinese** [comment]
Interestingly, the error still happens after logging out. Only a Clear Data (and remove /AnkiDroid folder), followed by import deck, will the error not occur.

**Comment by mikehardy** [comment]
Have you tried the alphas and the newest version of the API, and looked at the new sample code? I was just in there adding some defensive code, but I doubt that will actually be the fix. I'm not familiar enough with that area but I'd start by looking at a dB dump of a fresh install before and after sync with an empty account...

**Comment by timrae** [comment]
Yeah, we don't allow any actions in the API which would force a full sync since we want the user to explicitly agree to that, which is hard to do between apps

**Comment by mikehardy** [comment]
Seems sensible. Not that I want to sign up for work by proposing something but is it possible for the content provider API to generate pop-ups? If so we could get anki-side consent and advise them. If not then it seems like the error could be clearer.. . No way I would have guessed this...

**Comment by timrae** [comment]
I'm not sure, I put it in the too hard bucket at the time and that hasn't changed, but if someone figures out a way to handle it without requiring us to trust client applications to handle it properly then I'll be open to discuss it

**Comment by mikehardy** [comment]
Just to make sure I'm thinking about things correctly, would a successful attempt be a flow like:
1- someone asks ContentProvider to do something that would require force-sync
2- instead of us throwing an exception here, we pop up a dialog explaining the situation, the consequences (and maybe saving that they're using the API for later ACRA reporting...) and asking approval
3- if they don't approve then we return error from the API (null or -1 or whichever, which we should be doing now anyway in place of this exception)
4- if they do approve we do the thing?

No idea if 2 is possible, 3 at least may be something you could do @hanpingchinese to improve the current situation at least...

**Comment by hanpingchinese** [comment]
Thanks for the replies everyone.

Just so I understand, the reason the API allows this when not logged in is because it does not require a full sync?  But then why does the problem persist after logging out?

I'm wondering if there are any instructions I can pass on to the user to allow this card template insert.  at the moment the steps I have are:

1. Export all AnkiDroid data
2. Clear Data on AnkiDroid app and remove /AnkiDroid folder (though not /AnkiDroid/export !)
3. Launch AnkiDroid, but don't login
4. Import AnkiDroid using export from step 1
5. Use my app to insert the card template (it should work this time)
6. Login and sync (using "upload" option if prompted).

**Comment by timrae** [comment]
> But then why does the problem persist after logging out?

The logic is in here:
https://github.com/ankidroid/Anki-Android/blob/d98f676d094d5af8c89ad7d7f84916c4bbaf10e6/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java#L403

Basically the way it works is that any change which takes the state of the collection from "not needing a full sync" to "needing a full sync" needs to be explicitly approved by the user. After that point, changes don't need to be approved, because a full sync is already necessary. For a brand-new collection that's never been synced before, it already starts off in the state of "needing a full sync" (i.e. `ls` (last sync) value is set to zero, so `schemaChanged()` is always true). Logging out has no effect on this.

> I'm wondering if there are any instructions I can pass on to the user to allow this card template insert.

It's a bit involved, but basically they have to do this:

1. Sync in AnkiDroid to push changes to AnkiWeb, then sync in Anki Desktop to push changes to AnkiWeb, and finally sync again in AnkiDroid to pull all changes from AnkiWeb.
1. AnkiDroid and Anki Desktop should now be totally in sync. Manually verify this, checking that recent changes made on Anki Desktop are visible in AnkiDroid and vice-versa.
1. In AnkiDroid go Settings->Advanced->Force full sync. An "ok" message should briefly display.
1. Use Hangping app to add new card template (it should work this time)
1. Sync in AnkiDroid, choosing "upload to ankiweb" when prompted
1. Sync in Anki Desktop choosing "download from ankiweb" when prompted
1. AnkiDroid and Anki Desktop should now be totally in sync again, and both include the new card template in the Hangping model. Manually verify this in both Anki Desktop and AnkiDroid.

**Comment by hanpingchinese** [comment]
Excellent, thanks Tim

> For a brand-new collection that's never been synced before, it already starts off in the state of "needing a full sync"

So if the state is "needing a full sync" then the insert-card-template-via-api will work. But surely, after step 3 (force full sync) the collection is not in that state?

**Comment by timrae** [comment]
The subtitle for that option is "on next sync, forces changes in one direction". It doesn't actually do the sync, rather just forces it to the state "needs full sync"

**Comment by timrae** [comment]
I should also note that these instructions are assuming they have exactly one Android device and one PC. If the user has more than two clients, then of course they need to sync from each of them before and after the full sync.

**Comment by hanpingchinese** [comment]
Sure, my app already creates new note types, it's just that now is the first time I wanted to add a new card type to an existing note type.  I guess this would be a fair summary of the steps:

1. Make sure all Anki client data is synced to AnkiWeb server
2. AnkiDroid Settings > Advanced > Force full sync
3. 3rd party app inserts new card template (for existing note type) via API
4. AnkiDroid sync (upload to ankiweb when prompted)

> 2- instead of us throwing an exception here, we pop up a dialog explaining the situation, the consequences (and maybe saving that they're using the API for later ACRA reporting...) and asking approval

Since inserting a card type for an existing note type requires user permission (AnkiDroid UI), you could add support for an activity intent, that accepts extras: NOTE_TYPE_ID, CARD_NAME, QFMT, AFMT

**Comment by timrae** [comment]
It's important for the user to do this as well:

5. Repeat step 1

Otherwise the next time they sync on those other clients they'll lose the progress as a full sync download from AnkiWeb will be required.

**Comment by timrae** [comment]
And yes, we could make an intent to allow apps to request a full sync

**Comment by hanpingchinese** [comment]
Yes, and I think it would be better to have a more specific exception than IllegalArgumentException("Unable to add template") to better distinguish request-full-sync issues from other ones.

**Comment by timrae** [comment]
Feel free to make a PR

**Comment by marcardar** [comment]
I'm a bit too busy these days unfortunately, so feel free, anyone else, to give it a go!

**Comment by hanpingchinese** [comment]
I'm a bit too busy these days unfortunately, so feel free, anyone else, to give it a go!

**Comment by mikehardy** [comment]
I don't have the time to do the whole Intent thing, but I can add info to the exception so API consumers will find this when they search...
